src/compiler/interface.jl

```diff
@@ -171,6 +173,39 @@ function copy!(x::AbstractVector,  gs::Grads)
   x
 end
 
+# # Gradient Algebra: unary 
+# for op in (:+, :-)
+#   @eval broadcasted(::typeof($op), gs::Grads) = map($op, gs)
+# end
+
+# # Gradient Algebra: binary 
+# for op in (:+, :-)
+#   @eval broadcasted(::typeof($op), gs1::Grads, gs2::Grads) = map($op, gs1, gs2)
+# end
+
+broadcasted(f, gss::Grads...) = map(f, gss...)
+broadcasted(f, gs1::Grads, xs...) = map(f, gs1, xs...)
+broadcasted(::typeof(*), a::Number, gs::Grads) = map(*, gs, a)
+
+function Base.map(f, gs1::Grads, gss::Grads...)
+  @assert all(issetequal(gs1.params, gs.params) for gs in gss)
+  grads = IdDict{Any,Any}()
+  ps = Params(gs1.params)
+  for p in ps
+    grads[p] = f(gs1[p], (gs[p] for gs in gss)...) 
+  end
+  return Grads(grads, ps)
+end
+
+function Base.map(f, gs1::Grads, xs...)
```

Comment:
I'm not sure this is ok to do, I wonder if not specifying the type of `xs` could lead to ambiguity errors

Reply:
Yeah this seems unclear when `xs` contains some `Grads` with other types too.